### ISSUE TITLE
Remove a first-chance exception in Git blame

### ISSUE THREAD

**Issue by KirillOsenkov** [KirillOsenkov/SourceBrowser#58]
Currently when indexing any project there's a first-chance exception from libgit2 followed up by an exception log entry:

```
16:32:49 1/30/2017 4:32:49 PM: First chance exception while processing assembly: WpfApplication1
LibGit2Sharp.NotFoundException: the path 'obj' does not exist in the given tree
   at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) in c:\Git\libgit2sharp\LibGit2Sharp\Core\Ensure.cs:line 160
16:32:49 Couldn't blame C:\SourceBrowser\TestCode\WpfApplication1\obj\Debug\Page1.g.cs
16:32:49 Couldn't blame C:\SourceBrowser\TestCode\WpfApplication1\obj\Debug\GeneratedInternalTypeHelper.g.cs
LibGit2Sharp.NotFoundException: the path 'obj' does not exist in the given tree
   at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) in c:\Git\libgit2sharp\LibGit2Sharp\Core\Ensure.cs:line 160
   at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result) in c:\Git\libgit2sharp\LibGit2Sharp\Core\Ensure.cs:line 179
   at LibGit2Sharp.Core.Proxy.git_blame_file(RepositorySafeHandle repo, FilePath path, GitBlameOptions options) in c:\Git\libgit2sharp\LibGit2Sharp\Core\Proxy.cs:line 106
   at LibGit2Sharp.BlameHunkCollection..ctor(Repository repo, RepositorySafeHandle repoHandle, String path, BlameOptions options) in c:\Git\libgit2sharp\LibGit2Sharp\BlameHunkCollection.cs:line 45
   at LibGit2Sharp.Repository.Blame(String path, BlameOptions options) in c:\Git\libgit2sharp\LibGit2Sharp\Repository.cs:line 831
   at LibGit2Sharp.RepositoryExtensions.Blame(IRepository repository, String path) in c:\Git\libgit2sharp\LibGit2Sharp\RepositoryExtensions.cs:line 631
   at GitGlyph.GitBlameVisitor.GetBlame(String path) in C:\SourceBrowser\src\GitGlyph\GitBlameVisitor.cs:line 88
16:32:49 Couldn't blame C:\SourceBrowser\TestCode\WpfApplication1\obj\Debug\GeneratedInternalTypeHelper.g.cs
```

It would be nice to see if libgit2 allows a way to probe for a given file whether it's a part of the repo and don't try to annotate it if not, so that we can avoid the exception.

If that's not possible, Plan B is to whitelist the first-chance exception so that it doesn't get logged.

**Comment by KirillOsenkov** [comment]
@ddurschlag this one as well

**Comment by ddurschlag** [comment]
This one is tricky -- a file that has been created/deleted repeatedly over time might not exist in any given tree, and that seems fine. In my opinion, the cleanest solution is to try/catch that exception around the Repository.Blame call. libgit2sharp does not support doing anything but throwing in this scenario.

**Comment by KirillOsenkov** [comment]
You already try/catch it, so that's fine. But two things are happening: the FirstChanceExceptionHandler is reporting it (and we should probably add it to the whitelist to ignore it there). Additionally you call Log.Error and that prints out the entire callstack of the exception. I'm thinking of just calling Log.Message with just a single line, and use a different subdued console color for it. I think I should have a method for logging a message with a custom console color already.

**Comment by KirillOsenkov** [comment]
Also the exception I saw was only for the files in obj\Debug, which are not tracked by Git (they're excluded via .gitignore). I'm wondering if libgit2 lets you easily check if a file is being ignored and if so, don't run your feature for that file.

**Comment by ddurschlag** [comment]
I see, I forgot you had first-chance handling on. Turning the error into a less noisy level sounds right, as does whitelisting the first-chance handling. As to ignores: https://github.com/libgit2/libgit2sharp/blob/master/LibGit2Sharp.Tests/IgnoreFixture.cs#L39 . Looks like that's a one-liner.

**Comment by KirillOsenkov** [comment]
Oh, great! I thought they must have support for it.

**Comment by ddurschlag** [comment]
Interestingly, we can't check if the exception is a LibGit2Sharp.NotFoundException without taking a dependency on LibGit2Sharp outside the git plugin, which is kinda undesirable (or doing some nasty string checks). I wonder if we'd be better off with a clever solution to detecting exceptions from plugins in general and ignoring them.

**Comment by KirillOsenkov** [comment]
You can define IFirstChanceExceptionHandler with bool ShouldIgnore(Exception ex), export one from your extension and [ImportMany] them from the central handler.

**Comment by ddurschlag** [comment]
Wouldn't that allow plugins to accidentally handle each other's (or the core's) exceptions? I think investigating the stack trace to programmatically check if something came from a plugin would be more universal (I can't think of a situation where a plugin exception should take anything down) and safer.

**Comment by KirillOsenkov** [comment]
Well each handler should handle not only by the exception type but also analysing the ex.ToString() and search for a substring that would be unique enough. 

See here for examples:
http://sourcebrowser.azurewebsites.net/#HtmlGenerator/Utilities/FirstChanceExceptionHandler.cs,35

**Comment by ddurschlag** [comment]
@KirillOsenkov ,

Do you have a simple repro for this so I can confirm my solution?

**Comment by KirillOsenkov** [comment]
I found that just running HtmlGenerator and passing it the path to TestSolution.sln (`..\..\..\TestCode\TestSolution.sln`) will trigger it.

**Comment by ddurschlag** [comment]
Is this now resolved with #60 ?

**Comment by KirillOsenkov** [comment]
Yes, thank you.
